docs/reference/troubleshooting/common-issues/mapping-explosion.asciidoc

```diff
@@ -41,13 +41,16 @@ timing out in the browser's Developer Tools Network tab.
 doesn't normally cause problems unless it's combined with overriding
 <<mapping-settings-limit,`index.mapping.total_fields.limit`>>. The
 default `1000` limit is considered generous, though overriding to `10000`
-doesn't cause noticable impact depending on use case. However, to give
+doesn't cause noticeable impact depending on use case. However, to give
 a bad example, overriding to `100000` and this limit being hit
 by mapping totals would usually have strong performance implications.
 
 If your index mapped fields expect to contain a large, arbitrary set of
 keys, you may instead consider:
 
+* Setting <<mapping-settings-limit,`index.mapping.total_fields.ignore_dynamic_beyond_limit`>> to `true`.
```

Comment:
I have some half-baked thoughts about introducing a new dynamic fields limit instead, with the semantics of what you are implementing. We talked before about applying total fields limit only on dynamic mapping updates only, but were not sure how that would work in practice with a mix of static and dynamic mappings updates. See https://github.com/elastic/elasticsearch/issues/89911#issuecomment-1241112161 .

I am thinking that this new semantics could be the answer to that problem, and would allow us to entirely address # 89911. With that I would consider deprecating the total fields limit in favour of a new dynamic fields limit that has different semantics where only dynamic fields contribute to the count and no documents are rejected. How would you feel about making this one day the default behaviour of Elasticsearch?

This would address a couple of problems:
- avoid breaking changes around changing the behaviour of total fields limit (if we went and applied it only to dynamic updates
- introduce new semantics around how the total fields limit is applied, that removes the nuisance of the total fields limit for users while maintaining the safety around not allowing bogus documents to create thousands of dynamic fields
- simplify how users interact with the new semantics: I don't like the fact that an additional setting is required that works in combination with a pre-existing settings and changes its semantics when set. If we could have a single setting to make the switch independently of the other, that would be better.

I'd like to bring this for discussion with the team. What do you think?
We don't have to make such forward looking decisions in this PR, but if we see fit for this long term vision that will affect how we expose the new behaviour, which I think is important to get right from the get-go.

Reply:
I think in principle I agree that it makes sense to apply the limit only to dynamic mapping updates. But it also creates some complexity by us having to keep track of which fields have been added statically vs dynamically in the mapping so that we know how many fields have been added via dynamic mappings and how many can still be added.

Overall, I'm not sure if that's a problem that's worth solving. I agree it's a bit annoying to have to manually increase the field limit if a static mapping update fails when the update exceeds the limit. But I'd put this into the "paper cut" category, and I'm not sure if it's worth adding complexity and going through a deprecation cycle for.

While I don't see a need for it, are we sure that we don't need a mode where we reject documents with too many dynamic fields?

> I don't like the fact that an additional setting is required that works in combination with a pre-existing settings and changes its semantics when set.

Wouldn't we have a similar situation between the current `total_fields.limit` and a potentially new `dynamic_fields.limit` option?

It also doesn't seem trivial to completely remove `total_fields.limit`. Not sure how we could do that without a breaking change.

Reply:
I think if we merge the PR as is, we're still leaving the door open to do what you've proposed later on.

But again, I'm still skeptical whether that's something we'd even want to do. This would also make users have to think about how many fields have been added due to a dynamic or a static mapping update. They can't just count the total fields in the index anymore. They need to know how many dynamic vs static fields have been added in order to reason about why the field limit has been hit or how close they are to the field limit.

If you still think it's a path we want to go some day, I urge you to find a solution that doesn't delay this PR.

Reply:
> I think if we merge the PR as is, we're still leaving the door open to do what you've proposed later on.

Not really, because the new setting depends on the existing total fields limit. I'd like to consider that it is an independent new setting. This is the main reason for discussing this as part of this PR. It affects how we expose the new behavior we are introducing, and that is something that we won't be able to change easily once it's out.

I am not meaning to delay this PR. As I said, i will bring this to the team for discussion and we'll make a decision about whether we go for the boolean flag or a new total fields limit with the new semantics. I don't foresee additional changes to be made in this PR besides how the behaviour is exposed.

Reply:
> I think in principle I agree that it makes sense to apply the limit only to dynamic mapping updates. But it also creates some complexity by us having to keep track of which fields have been added statically vs dynamically in the mapping so that we know how many fields have been added via dynamic mappings and how many can still be added.

Indeed, that is what I am considering to avoid. With the behaviour you are introducing, I wonder if we need to adapt the current total fields limit to only apply to dynamic fields limit. We could say the new way is the best way, it applies only to dynamic fields and does not reject documents. I don't mean that we need to decide all of this right now. But if we think that's a possibility we should try and make the new setting independent of the existing one.

Reply:
I think I understand now. So you're saying with that new field limit setting, we'd basically not validate the field limit for static mapping updates, only for dynamic updates. But the limit still applies to the total number of fields and we don't need to remember which fields have been added statically or dynamically.

Let me make an example to see if we're on the same page
* The limit is set to 1000
* User adds 999 fields via a static mapping update
* User indexes a document that would add 2 new fields dynamically
* We'll add one field to the mappings and ignore the other field; the document will not be rejected.

Does that sound about right?

When it comes to removing the current field limit in favor of the new one, we should consider still reading out the old setting and applying it to the new one. Also, what would be the behavior if both the old and the new limit setting is configured?

This ambiguity and compatibility story makes me wonder whether some kind of boolean flag wouldn't make our lives easier. We could also flip the default value of that boolean at some point, deprecate and finally remove the flag.

Reply:
I brought this topic for discussion with the team. Our conclusion is that we are fine with the current boolean flag. Whatever change we want to make in the future, either that be changing the semantics of the current total fields limit or adding a new limit, there is an upgrade path possible. It is also very difficult to foresee exactly what type of changes we will make in the future. This new semantics is promising and we should consider making it default one day perhaps.

Two questions for confirmation:

1. the new semantics are applied to any dynamic field, including those that are the result of applying a dynamic template and not directly provided by users in documents? That's fine, but I wanted to verify my understanding.

2. with the new semantics, we still apply the total fields limit as-is to put mappings calls. Would we want to change that perhaps too in the future and no longer limit the number of fields that can be manually mapped when the new behavior is turned on?

Reply:
> the new semantics are applied to any dynamic field, including those that are the result of applying a dynamic template and not directly provided by users in documents?

Not entirely sure I understand. Aren't all dynamic mapping updates a result of an index operation for a document that has fields that aren't in the mappings already? Whether or not the new fields match an index template just has an effects whether a default mapping that's based on the dynamic type is used or a more customized one.

> Would we want to change that perhaps too in the future and no longer limit the number of fields that can be manually mapped when the new behavior is turned on?

Reply:
I mean that you may have dynamic templates defined in the mappings that affect how dynamic fields are mapped, and may cause mapping additional fields that are not originally present in incoming documents.

> The setting is called ignore_dynamic_beyond_limit so it might be surprising if the flag also affects how static mappings are handled.

I can see that, but I also take that you don't care that much about static mappings and the majority of your fields are going to come in as dynamic fields as opposed to defined via put mapping/create index. I initially thought it may be surprising that put mappings hits the fields limit once ignore_dynamic_beyond_limit is on, but maybe it's not. 

It came up with the team that the limit should not be applied to static mappings, as users that hit it that way will just go and increase the limit which is kind of pointless. But this is more about the semantics of the total fields limit than about the mechanics of the new behavior. We are free to adjust things further but it's not required in this PR.

The fact that you mention the need for another setting gives me second thoughts, but let's get this one in :)

Reply:
> I mean that you may have dynamic templates defined in the mappings that affect how dynamic fields are mapped, and may cause mapping additional fields that are not originally present in incoming documents.

Gotcha, all of those are covered with the flag as well.